public/app/core/history/richHistoryLocalStorage.ts

```diff
@@ -0,0 +1,172 @@
+    store.setObject(RICH_HISTORY_KEY, updatedHistory);
+  }
+
+  async updateStarred(id: number, starred: boolean): Promise<void> {
+    const richHistory: RichHistoryQuery[] = store.getObject(RICH_HISTORY_KEY, []);
+    const updatedHistory = richHistory.map((query) => {
+      if (query.ts === id) {
+        query.starred = starred;
+      }
+      return query;
+    });
+
+    store.setObject(RICH_HISTORY_KEY, updatedHistory);
+  }
+
+  async updateComment(id: number, comment: string): Promise<void> {
+    const richHistory: RichHistoryQuery[] = store.getObject(RICH_HISTORY_KEY, []);
+    const updatedHistory = richHistory.map((query) => {
+      if (query.ts === id) {
+        query.comment = comment;
+      }
+      return query;
+    });
+    store.setObject(RICH_HISTORY_KEY, updatedHistory);
+  }
+}
+
+/**
+ * Removes entries that do not match retention policy criteria.
+ */
+function cleanUp(richHistory: RichHistoryQuery[]): RichHistoryQuery[] {
+  const retentionPeriod: number = store.getObject(RICH_HISTORY_SETTING_KEYS.retentionPeriod, 7);
+  const retentionPeriodLastTs = createRetentionPeriodBoundary(retentionPeriod, false);
+
+  /* Keep only queries, that are within the selected retention period or that are starred.
+   * If no queries, initialize with empty array
+   */
+  return richHistory.filter((q) => q.ts > retentionPeriodLastTs || q.starred === true) || [];
+}
+
+/**
+ * Ensures the entry can be added. Throws an error if current limit has been hit.
+ * Returns queries that should be saved back giving space for one extra query.
+ */
+function checkLimits(queriesToKeep: RichHistoryQuery[]): { queriesToKeep: RichHistoryQuery[]; limitExceeded: boolean } {
+  // remove oldest non-starred items to give space for the recent query
+  let limitExceeded = false;
+  let current = queriesToKeep.length - 1;
+  while (current >= 0 && queriesToKeep.length >= MAX_HISTORY_ITEMS) {
+    if (!queriesToKeep[current].starred) {
+      queriesToKeep.splice(current, 1);
+      limitExceeded = true;
+    }
+    current--;
+  }
+
+  return { queriesToKeep, limitExceeded };
+}
+
+function migrateRichHistory(richHistory: RichHistoryQuery[]) {
+  const transformedRichHistory = richHistory.map((query) => {
+    const transformedQueries: DataQuery[] = query.queries.map((q, index) => createDataQuery(query, q, index));
+    return { ...query, queries: transformedQueries };
+  });
+
+  return transformedRichHistory;
+}
+
+function createDataQuery(query: RichHistoryQuery, individualQuery: DataQuery | string, index: number) {
+  const letters = 'ABCDEFGHIJKLMNOPQRSTUVXYZ';
+  if (typeof individualQuery === 'object') {
+    // the current format
+    return individualQuery;
+  } else if (isParsable(individualQuery)) {
+    // ElasticSearch (maybe other datasoures too) before grafana7
+    return JSON.parse(individualQuery);
+  }
+  // prometehus (maybe other datasources too) before grafana7
+  return { expr: individualQuery, refId: letters[index] };
```

Comment:
question: `individualQuery` at this point "should" be a string, how would `if (isParsable(individualQuery))` ever fail?

Reply:
`isParsable` may return false if the string is not a valid JSON. We can remove this code once we migrate to backend data sources.

Reply:
whooops, i misread the function!